src/of_type.jl

```diff
@@ -0,0 +1,1602 @@
+    end
+
+    names = get_names(T)
+    types = get_types(T)
+
+    # Separate constants from values
+    constants = Dict{Symbol,Any}()
+    values = Dict{Symbol,Any}()
+
+    for (key, val) in pairs(kwargs)
+        idx = findfirst(==(key), names)
+        if idx !== nothing && types.parameters[idx] <: OfConstantWrapper
+            constants[key] = val
+        else
+            values[key] = val
+        end
+    end
+
+    # Check that all constants are provided
+    for (idx, name) in enumerate(names)
+        if types.parameters[idx] <: OfConstantWrapper && !haskey(constants, name)
+            error("Constant `$name` is required but not provided")
+        end
+    end
+
+    # First concretize with constants
+    concrete_type = of(T, NamedTuple(constants))
+
+    # Check if all constants are resolved
+    if has_symbolic_dims(concrete_type)
+        missing_symbols = get_unresolved_symbols(concrete_type)
+        error("Missing values for symbolic dimensions: $(join(missing_symbols, ", "))")
+    end
+
+    # Get the names and types from the concrete type (constants removed)
+    concrete_names = get_names(concrete_type)
+    concrete_types = get_types(concrete_type)
+
+    # Build the result with provided values or defaults
+    result_values = Any[]
+    for (idx, name) in enumerate(concrete_names)
+        field_type = concrete_types.parameters[idx]
+
+        if haskey(values, name)
+            # Validate the provided value
+            try
+                push!(result_values, _validate(field_type, values[name]))
+            catch e
+                error("Validation failed for field $name: $(e.msg)")
+            end
+        else
+            # Use the value generator function
+            push!(result_values, value_generator(field_type))
+        end
+    end
+
+    # Return the instance as a NamedTuple
+    return NamedTuple{concrete_names}(Tuple(result_values))
+end
+
+# ========================================================================
+# Parameterized Constructor with Validation
+# ========================================================================
+function (::Type{T})(; kwargs...) where {T<:OfType}
+    return _create_instance_impl(T, zero, kwargs)
+end
+
+# Constructor with default_value as positional argument
+function (::Type{T})(default_value; kwargs...) where {T<:OfType}
+    return _create_instance_impl(
+        T, field_type -> _create_with_default(field_type, default_value), kwargs
+    )
+end
+
+# ========================================================================
+# @of Macro and Related Processing Functions
+# ========================================================================
+
+"""
+    @of(field1=spec1, field2=spec2, ...)
+
+Create an OfNamedTuple type with cleaner syntax that allows direct references
+between fields without needing symbols.
+
+# Examples
+```julia
+T = @of(
+    rows=of(Int; constant=true),
+    cols=of(Int; constant=true),
+    data=of(Array, rows, cols)
+)
```

Comment:
Does order matter here? Could `data` come before `rows` and `cols`? I assume `rows` and `cols` have to be `constant=true` for them to be used in creating `data`? Would be good to document these, and then probably stick the doctsring of `@of` and `of` in the docs as well.